nixos/modules/services/security/bitwarden_rs/default.nix

```diff
@@ -0,0 +1,130 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.bitwarden_rs;
+  user = config.users.users.bitwarden_rs.name;
+  group = config.users.groups.bitwarden_rs.name;
+
+  # Convert name from camel case (e.g. disable2FARemember) to upper case snake case (e.g. DISABLE_2FA_REMEMBER).
+  nameToKey = name:
+    let
+      parts = builtins.split "([A-Z0-9]+)" name;
+      partsToKey = parts: removePrefix "_" (foldl' (key: x:
+        if isList x then key + "_" + head x
+        else if key != "" && elem (substring 0 1 x) lowerChars then # to handle e.g. [ "disable" [ "2FAR" ] "emember" ]
+          let last = stringLength key - 1;
+          in substring 0 last key + optionalString (substring (last - 1) 1 key != "_") "_" + substring last 1 key + toUpper x
+        else key + toUpper x) "" parts);
+    in partsToKey parts;
+
+  configFile = pkgs.writeText "bitwarden_rs.env" (concatMapStrings (s: s + "\n") (
+    (concatLists (mapAttrsToList (name: value:
+      if value != null then [ "${nameToKey name}=${if isBool value then boolToString value else toString value}" ] else []
+    ) cfg.config))));
+
+in {
+  options.services.bitwarden_rs = with types; {
+    enable = mkEnableOption "bitwarden_rs";
+
+    backupDir = mkOption {
+      type = nullOr str;
+      default = null;
+      description = ''
+        The directory under which bitwarden_rs will backup its persistent data.
+      '';
+    };
+
+    config = mkOption {
+      type = attrsOf (nullOr (either (either bool int) str));
+      default = {};
+      example = literalExample ''
+        {
+          signupsAllowed = true;
+          domain = https://example.com;
+          rocketPort = 8222;
+          rocketLog = "critical";
+        }
+      '';
+      description = ''
+        The configuration of bitwarden_rs is done through environment variables,
+        therefore the names are converted from camel case (e.g. disable2FARemember)
+        to upper case snake case (e.g. DISABLE_2FA_REMEMBER).
+        In this conversion digits (0-9) are handled just like upper case characters,
+        so foo2 would be converted to FOO_2.
+
+        Based on the attributes passed to this config option a environment file will be generated
+        that is passed to bitwarden_rs's systemd service.
+
+        The available configuration options can be found in the environment template file:
+        https://github.com/dani-garcia/bitwarden_rs/blob/1.8.0/.env.template
+      '';
+      apply = config: let cfg = {
+        dataFolder = "/var/lib/bitwarden_rs";
+        webVaultEnabled = true;
+      } // config; in cfg // optionalAttrs (cfg.webVaultEnabled && !(cfg ? webVaultFolder)) {
```

Comment:
I tried some other things, but this was the only way I got it to work, so if you have some better way of doing this, I very much welcome it.

Reply:
Does it matter that `webVaultFolder` is set even when it's not used? Because if it doesn't matter (aka the program just ignores it), you could unconditionally set it.

Reply:
In that case this is what I'd do:
```diff
diff --git a/nixos/modules/services/security/bitwarden_rs/default.nix b/nixos/modules/services/security/bitwarden_rs/default.nix
index d41b6e0dbd1..0b9d6037cf1 100644
--- a/nixos/modules/services/security/bitwarden_rs/default.nix
+++ b/nixos/modules/services/security/bitwarden_rs/default.nix
@@ -60,22 +60,13 @@ in {
         The available configuration options can be found in the environment template file:
         https://github.com/dani-garcia/bitwarden_rs/blob/1.8.0/.env.template
       '';
-      apply = config: let cfg = {
-        dataFolder = "/var/lib/bitwarden_rs";
-        webVaultEnabled = true;
-      } // config; in cfg // optionalAttrs (cfg.webVaultEnabled && !(cfg ? webVaultFolder)) {
-        webVaultFolder = "${pkgs.bitwarden_rs-vault}/share/bitwarden_rs/vault";
-      };
     };
   };
 
   config = mkIf cfg.enable {
-    assertions = singleton {
-      assertion = cfg.config.dataFolder == "/var/lib/bitwarden_rs";
-      message = ''
-        Due to the way bitwarden_rs's systemd service has been setup it is not possible to change the data folder
-        to something other than the default location (/var/lib/bitwarden_rs).
-      '';
+    services.bitwarden_rs.config = {
+      dataFolder = "/var/lib/bitwarden_rs";
+      webVaultFolder = "${pkgs.bitwarden_rs-vault}/share/bitwarden_rs/vault";
     };
 
     users.users.bitwarden_rs = { inherit group; };
```

I don't think the assertion is necessary, because the user would have to use `mkForce` to set it, at which point they should realized that they need to change the module to have their special value work.

Reply:
The downside of always setting `webVaultFolder` is that it would cause it to be installed even when not being used.

Reply:
Ah I see..

Reply:
Alright I think apply can indeed not be gotten rid of, but I'd still put whatever you can in the `config` section, then the module system can handle the merging:

```diff
diff --git a/nixos/modules/services/security/bitwarden_rs/default.nix b/nixos/modules/services/security/bitwarden_rs/default.nix
index c0882711cd6..df64f38631d 100644
--- a/nixos/modules/services/security/bitwarden_rs/default.nix
+++ b/nixos/modules/services/security/bitwarden_rs/default.nix
@@ -69,24 +69,17 @@ in {
         The available configuration options can be found in
         <link xlink:href="https://github.com/dani-garcia/bitwarden_rs/blob/1.8.0/.env.template">the environment template file</link>.
       '';
-      apply = config: let cfg = {
-        dataFolder = "/var/lib/bitwarden_rs";
-        webVaultEnabled = true;
-      } // config // {
-        inherit (cfg) domain;
-      }; in cfg // optionalAttrs (cfg.webVaultEnabled && !(cfg ? webVaultFolder)) {
+      apply = config: optionalAttrs config.webVaultEnabled {
         webVaultFolder = "${pkgs.bitwarden_rs-vault}/share/bitwarden_rs/vault";
-      };
+      } // config;
     };
   };
 
   config = mkIf cfg.enable {
-    assertions = singleton {
-      assertion = cfg.config.dataFolder == "/var/lib/bitwarden_rs";
-      message = ''
-        Due to the way bitwarden_rs's systemd service has been setup it is not possible to change the data folder
-        to something other than the default location (/var/lib/bitwarden_rs).
-      '';
+    services.bitwarden_rs.config = {
+      dataFolder = "/var/lib/bitwarden_rs";
+      webVaultEnabled = mkDefault true;
+      domain = mkDefault cfg.domain;
     };
 
     users.users.bitwarden_rs = { inherit group; };
```

I also made the `apply` a bit simpler by switching the order of the arguments to `//` (then `webVaultFolder` from config overrides the default one). Feel free to keep the assertion in if you want, I personally don't think it's needed.